// @saberhq/anchor-contrib/dist/esm/utils/provider.js
import { Provider as AnchorProvider } from "@project-serum/anchor";
import { SolanaProvider } from "@saberhq/solana-contrib";
/**
 * Creates a readonly Saber Provider from an Anchor provider.
 * @param anchorProvider The Anchor provider.
 * @returns
 */
export const makeReadonlySaberProvider = (anchorProvider) => {
    return SolanaProvider.load({
        connection: anchorProvider.connection,
        wallet: anchorProvider.wallet,
        opts: anchorProvider.opts,
    });
};
/**
 * Creates a Saber Provider from an Anchor provider.
 * @param anchorProvider The Anchor provider.
 * @returns
 */
export const makeSaberProvider = (anchorProvider) => {
    return SolanaProvider.load({
        connection: anchorProvider.connection,
        wallet: anchorProvider.wallet,
        opts: anchorProvider.opts,
    });
};
/**
 * Creates an Anchor Provider from a Saber provider.
 * @param saberProvider
 * @returns
 */
export const makeAnchorProvider = (saberProvider) => {
    return new AnchorProvider(saberProvider.connection, saberProvider.wallet, saberProvider.opts);
};
