// @saberhq/anchor-contrib/dist/cjs/utils/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAnchorProvider = exports.makeSaberProvider = exports.makeReadonlySaberProvider = void 0;
const anchor_1 = require("@project-serum/anchor");
const solana_contrib_1 = require("@saberhq/solana-contrib");
/**
 * Creates a readonly Saber Provider from an Anchor provider.
 * @param anchorProvider The Anchor provider.
 * @returns
 */
const makeReadonlySaberProvider = (anchorProvider) => {
    return solana_contrib_1.SolanaProvider.load({
        connection: anchorProvider.connection,
        wallet: anchorProvider.wallet,
        opts: anchorProvider.opts,
    });
};
exports.makeReadonlySaberProvider = makeReadonlySaberProvider;
/**
 * Creates a Saber Provider from an Anchor provider.
 * @param anchorProvider The Anchor provider.
 * @returns
 */
const makeSaberProvider = (anchorProvider) => {
    return solana_contrib_1.SolanaProvider.load({
        connection: anchorProvider.connection,
        wallet: anchorProvider.wallet,
        opts: anchorProvider.opts,
    });
};
exports.makeSaberProvider = makeSaberProvider;
/**
 * Creates an Anchor Provider from a Saber provider.
 * @param saberProvider
 * @returns
 */
const makeAnchorProvider = (saberProvider) => {
    return new anchor_1.Provider(saberProvider.connection, saberProvider.wallet, saberProvider.opts);
};
exports.makeAnchorProvider = makeAnchorProvider;
